packages/manager/modules/pci/src/projects/project/kubernetes/details/restrictions/controller.js

```diff
@@ -0,0 +1,112 @@
+import IP from './IP.class';
+
+export default class KubernetesRestrictionsCtrl {
+  /* @ngInject */
+  constructor($translate, CucCloudMessage) {
+    this.$translate = $translate;
+    this.CucCloudMessage = CucCloudMessage;
+  }
+
+  $onInit() {
+    this.loadMessages();
+    this.getIps(this.restrictions);
+    this.selectRow(this.ips.length);
+    this.currentIp = null;
+  }
+
+  loadMessages() {
+    this.CucCloudMessage.unSubscribe(
+      'pci.projects.project.kubernetes.details.restrictions',
+    );
+    this.messageHandler = this.CucCloudMessage.subscribe(
+      'pci.projects.project.kubernetes.details.restrictions',
+      { onMessage: () => this.refreshMessages() },
+    );
+  }
+
+  refreshMessages() {
+    this.messages = this.messageHandler.getMessages();
+  }
+
+  getIps(restrictions) {
+    this.ips = restrictions.map((ip) => new IP(ip));
+  }
+
+  selectRow(index) {
+    this.currentEditedRow = index;
+  }
+
+  addRestriction() {
+    this.currentIp = null;
+    this.ips.push({});
+    this.selectRow(this.ips.length - 1);
+  }
+
+  editRestriction(rowIndex) {
+    this.selectRow(rowIndex);
+    this.currentIp = this.ips[rowIndex].address;
+    if (!this.ips[this.ips.length - 1].address) {
+      this.ips.pop();
+    }
+  }
+
+  deleteIp(ip, index) {
+    this.isLoading = true;
+    return this.deleteRestriction(ip.address)
+      .then(() => {
+        this.ips.splice(index, 1);
+        return this.CucCloudMessage.success(
+          this.$translate.instant('kube_restrictions_add_success'),
+        );
+      })
+      .catch(
+        (error) =>
+          `${this.CucCloudMessage.error(
+            this.$translate.instant('kube_restrictions_add_error'),
+          )}${error.data}`,
+      )
+      .finally(() => {
+        this.isLoading = false;
+      });
+  }
+
+  saveRestrictions(index) {
+    this.isLoading = true;
+    this.ips[index] = new IP(this.currentIp);
+    this.selectRow(this.ips.length);
+    return this.updateRestrictions(this.ips.map((ip) => ip.format()))
+      .then((restrictions) => {
+        this.getIps(restrictions);
+        return this.CucCloudMessage.success(
+          this.$translate.instant('kube_restrictions_add_success'),
+        );
+      })
+      .catch(
+        (error) =>
+          `${this.CucCloudMessage.error(
+            this.$translate.instant('kube_restrictions_add_error'),
+          )}${error.data}`,
+      )
+      .finally(() => {
+        this.isLoading = false;
+      });
+  }
+
+  isIPValid() {
```

Comment:
Why not having this method into the `IP` class ?

Reply:
I tried but if I use an array value such as `this.ips[$rowIndex]`, the input looses focus after each character (probably due to the way ngrepeat is handled by the datagrid tracking by $index https://github.com/angular/angular.js/issues/13327) so I have to workaround by using a different value which is not an IP instance

Reply:
Ok good to know